build.gradle

```diff
@@ -59,20 +59,103 @@ ext {
+                Math.min( GRADLE_MAX_SUPPORTED_BYTECODE_VERSION, javaVersions.test.compiler.asInt() ) )
+        javaVersions.test.launcher = javaVersions.test.compiler
+    }
+    else {
+        // Not testing a particular JDK version: we will use the same JDK used to run Gradle.
+        // We disable toolchains for convenience, so that anyone can just run the build with their own JDK
+        // without any additional options and without downloading the whole JDK.
+        javaToolchainEnabled = false
+        gradleJdkVersion = JavaLanguageVersion.of( JavaVersion.current().getMajorVersion() )
+        javaVersions = [
+                main: [
+                        compiler: gradleJdkVersion,
+                        release: baselineJavaVersion
+                ],
+                test: [
+                        compiler: gradleJdkVersion,
+                        release: baselineJavaVersion,
+                        launcher: gradleJdkVersion
+                ]
+        ]
+    }
+
+    println "Java versions for main code: " + project.javaVersions.main
+    println "Java versions for tests: " + project.javaVersions.test
 }
 
 subprojects {
     apply plugin: 'java-library'
     apply plugin: 'com.diffplug.gradle.spotless'
     group = 'org.hibernate.reactive'
     version = projectVersion
-    sourceCompatibility = project.baselineJavaVersion
-    targetCompatibility = project.baselineJavaVersion
-    compileJava.options.encoding = 'UTF-8'
+
+    if ( project.javaToolchainEnabled ) {
+        // "sourceCompatibility" is not supported with toolchains. We have to work around that limitation.
+        compileJava.configure {
+            if ( project.project.javaVersions.main.compiler.asInt() < 11 ) {
+                options.compilerArgs << '-source'
+                options.compilerArgs << project.javaVersions.main.release.toString()
+                options.compilerArgs << '-target'
+                options.compilerArgs << project.javaVersions.main.release.toString()
+            } else {
+                options.release = project.javaVersions.main.release.asInt()
+            }
+            options.encoding = 'UTF-8'
+        }
+        compileTestJava.configure {
+            if ( project.project.javaVersions.test.compiler.asInt() < 11 ) {
+                options.compilerArgs << '-source'
+                options.compilerArgs << project.javaVersions.test.release.toString()
+                options.compilerArgs << '-target'
+                options.compilerArgs << project.javaVersions.test.release.toString()
+            } else {
+                options.release = project.javaVersions.test.release.asInt()
+            }
+            options.encoding = 'UTF-8'
+        }
+    }
+    else {
+        sourceCompatibility = JavaVersion.toVersion( project.baselineJavaVersion )
+        targetCompatibility = JavaVersion.toVersion( project.baselineJavaVersion )
+        compileJava.options.encoding = 'UTF-8'
```

Comment:
Can't we just set this option once before the initial if?

Reply:
Right. I amended my last commit and force-pushed.